File: ruby/bob/bob.rb
class Remark
  def initialize(string)
    @string=string
  end

  def question?
    @string[-1] == "?"
  end

  def yelling?
    @string == @string.upcase && ('A'..'Z').any? { |f| @string.include?(f) }
  end

  def silence?
    @string.gsub(/\s/ , '') == ""
  end
end

class Bob
  def self.hey(string)
    answer Remark.new(string)
  end

  def self.answer(object)
    if object.yelling?
      'Whoa, chill out!'
    elsif object.question?
      'Sure.'
    elsif object.silence?
      'Fine. Be that way!'
    else
      'Whatever.'
    end
  end
end

File: elixir/bob/bob.exs
defmodule Bob do
  def hey(input) do
    new_input = String.strip(input)
    cond do
      is_empty?(new_input)    -> "Fine. Be that way!"
      is_question?(new_input) -> "Sure."
      is_yelling?(new_input)  -> "Whoa, chill out!"
      true                    -> "Whatever."
    end
  end


  defp is_question?(input) do
    input |> String.ends_with?("?")
  end

  def is_yelling?(input) do
    input |> String.upcase == input && String.match?(input, ~r/\p{L}/)
  end

  defp is_empty?(""), do: true
  defp is_empty?(_str), do: false
end
